lua: always run command_native_async callback asynchronously

CogentRedTester opened the pull request:
Previously if the raw command_native_async returned an error then the
callback function was run directly. This was at odds with the Javascript
implementation of the function, which would run the callback
asynchronously.

This commit replicates the Javascript behaviour in Lua using the
mp.add_timeout function.

This provides consistency in how the callback function is run, and
changes Lua and Javascript so that they behave the same.

I haven't actually tested this because I don't have the mpv build process
setup on my computer, but it is an extremely simple change and I
can't see how this wouldn't work.

CogentRedTester commented:
Perhaps someone more experienced than me could let me know if this is worthy of an addition to `interface-changes.rst`.

CogentRedTester commented:
> Did it bite you in practice? or is it only a ganeral improvement?

Yes I did run into this issue. I had a script that was running a number of subprocess commands concurrently and they overflowed the event queue. I was using Lua coroutines to suspend execution and then resume inside the callback, but if the callback is run synchronously then Lua throws an error since execution hasn't been suspended yet. This was quite unexpected since the subprocess command doesn't seem to fail otherwise, and the size of the event queue is not made transparent to the user (I think it's 1000 based on my analysis of the C code).

> While I don't disagree that it's nicer to both always callback async, and that lua/js behave the same, I don't think this specific argument is a valid one, because it's not documented to callback async also on failure - unlike the js docs of this API.

I'm not claiming this is a bug fix, I consider it a feature change. However the manual states:
>Like mp.command_native(), but the command is ran asynchronously (as far as possible), and upon completion, fn is called.
> ...
>If starting the command failed for some reason, nil, error is returned, and fn is called indicating failure, using the same error value.

Which to me seems ambiguous, one could reasonably interpret this as meaning fn is asynchronous even on errors.

> And maybe (in a new commit?) also make the cb argument optional. It's documented as optional, but in practice it's required currently (unlike in js where it's indeed optional).

In this PR or in a different one?

CogentRedTester commented:
Latest commit should have addressed your three points.

avih commented:
> but if the callback is run synchronously

It only runs synchronously if it fails to start, not in general. Such failure presumably does not depend on the state of mpv or anyting else. It's likely due to bad command name or too many arguments or some such - something which would always fail before it starts.

For instrance, the `subprocess` command with bad program name still callback a-sync. But If you try to run the non-exiting command "subprocess-XX" then it would callback syncoronously.

Did you encounter failure to start (synchronous callback) for a command (+arguments) which could sometimes callback a-sync? If yes, examples please?

If no, then you should have had no issue, because it's presumably a developer error, not runtime error which depends on runtime conditions. It's akin to a line of code which calls a function which doesn't exist - it would always error out when/if this line is reached.

> Yes I did run into this issue. I had a script that was running a number of subprocess commands concurrently and they overflowed the event queue. I was using Lua coroutines to suspend execution and then resume inside the callback, but if the callback is run synchronously then Lua throws an error since execution hasn't been suspended yet. This was quite unexpected since the subprocess command doesn't seem to fail otherwise, and the size of the event queue is not made transparent to the user (I think it's 1000 based on my analysis of the C code).

To be honest, while I couldn't quite follow what you described here, I think you either use a wrong approch (coroutines to address mpv queue overflow?), or you're asking to much. While mpv scripting can be very powerful, it does have limits...

Also, this may or may not apply to you, but timers (like what your patch makes of the sync error callback) are only called once the event queue is empty. So as long as there are mpv events to process (of which "queue overflow" is a subset) then timers are not called, and in your case the failure callback will not be called. Please do _not_ open a PR to change this behavior of script timers.

The docs look fine now, thanks, but the commit message seem to rationalize mainly for parity with JS, while your true reasoning is a different one. So please explain (briefly) why always a-sync is better/nicer/whatever, and then mention that JS does the same.

Preferably, if you have a simple real-world example of a command which can sometimes callback synchronously and sometimes callback a-synchronously, do add it to the commit message too.

> In this PR or in a different one?

Here is fine, as long as it's at a separate commit. But I do really want to first understand the cases where the same command can sometimes callback synchronously and sometimes callback a-sync.

CogentRedTester commented:
> To be honest, while I couldn't quite follow what you described here, I think you either use a wrong approch (coroutines to address mpv queue overflow?), or you're asking to much. While mpv scripting can be very powerful, it does have limits...

Yes the use of coroutines is not important here, I was just explaining how I came across the error.

> Did you encounter failure to start (synchronous callback) for a command (+arguments) which could sometimes callback a-sync? If yes, examples please?
>
> If no, then you should have had no issue, because it's presumably a developer error, not runtime error which depends on runtime conditions. It's akin to a line of code which calls a function which doesn't exist - it would always error out when/if this line is reached.

Basically if you try to run more than 1000 `command_native_async` operations at once you'll get a queue overflow error even if the arguments are otherwise guaranteed to be successful. I believe this value is set in [this](https://github.com/mpv-player/mpv/blob/24f4582b6f7f57f566418551f9252b8578d2b602/player/client.c#L285) line. This extremely basic script demonstrates this:

```lua
local mp = require "mp"

for i = 1, 1001, 1 do
    local res, err = mp.command_native_async({
        name = "subprocess",
        args = {"cmd"},
        capture_stdout = true,
        playback_only = false
    }, function(success, result, err)
        if not success then print(result, err) end
    end)
end
```
You can change `1001` to `1000` to test this out.

This may seem unlikely to happen in practice, but if you're running these commands automatically on a large list the user may accidentally go over this number. For example the [mpv-playlistmanager](https://github.com/jonniek/mpv-playlistmanager) script does this automatically to get the titles of youtube videos; if the user is opening a playlist with more than 1000 items that's a non async callback that the script writer did not know to expect.

In addition, while doing research for this reply it turns out that this 1000 item queue limit seems to apply across all scripts, not just one. This means if multiple scripts make lots of these async calls they could interfere with each other. You can test this by adding the two copies of the following script:

```lua
local mp = require "mp"

local counter = 0
for i = 1, 1000, 1 do
    local res, err = mp.command_native_async({
        name = "subprocess",
        args = {"cmd"},
        capture_stdout = true,
        playback_only = false
    }, function()
        counter = counter + 1
    end)

    if not res then print(res, err) end
end

mp.add_timeout(5, function()
    print(counter)
end)
```

Interestingly this doesn't throw an error when the command_native_async command is run, but it does throw an mpv error in the console and ends up completely dropping many of the callback functions despite reporting a success. This is actually a completely separate issue to what I was talking about, so I may make another issue request about this at some point.

avih commented:
Oops, didn't mean to close.

CogentRedTester commented:
> But again, IMO you're not using the right tool for the job if you're trying to launch thousands of subprocesses from mpv scripts.

Oh I completely agree, and I'm not trying to change these limits, I'm just explaining some of the unexpected errors I have come across.

I'm mainly suggesting this patch for consistency: knowing that a callback function will always be run asynchronously just takes some weight off the developer. It means that if they don't know about the queue limit and they write their code assuming the callback will be asynchronous it has less chance to break things. It means if they take input from the user and it results in invalid arguments, or if they have some edge bugs with invalid data types then they may not need to write extra logic in the callback to handle synchronous and asynchronous execution.

Plus it brings things more in line with Javascript, which adds consistency across the different scripting APIs.

If you think this is decent reasoning I'll rewrite the commit message to say this.

CogentRedTester commented:
Ok I have rewritten the commit message, and have also added a new commit that changed the callback to be optional. I've renamed the PR as well to reflect both changes, but please feel free to rename it.

avih commented:
Thanks.

Pushed as 099ae8671732a333216a622cba7751f08deff9af (with minor title change) and 1ffdb9128d87dffe869c1e9f4024a16bfd0022c4 (minor title and message change).
